### PR TITLE
Rename Final to Unassignable

### PR DESCRIPTION
**Pull request by JackStouffer**
Now that https://github.com/dlang/phobos/pull/4755 was pulled this discussion can be had.

In the original PR https://github.com/dlang/phobos/pull/3862 there was agreement among many contributors that `Final` was a bad name, including myself, @ZombineDev, and @ntrel.

My main argument against `Final` is that things in the standard library should avoid having the same names as keywords. This is especially true when that symbol and the keyword do entirely different things. Keeping the name as is will only lead to confusion IMO.

HeadConst was also proposed to draw a parallel to C++, but this works different than C++ as the fields of structs are still assignable.

Ping @wilzbach @andralex

### CODE REVIEW

**Comment by andralex** [comment]
My personal preference goes for `Final`. Describing something by what it can't do is okay but most often if you can't do better. I think there's good precedent in other languages and in D itself to instantly make `Final` recognizable.

**Comment by AndrejMitrovic** [comment]
I like how terse `Final` is, it makes it convenient to use. I'm with Andrei on this one.

**Comment by DmitryOlshansky** [comment]
I was pro `Final` to begin with

**Comment by JackStouffer** [comment]
The people against are not core team members, so I will close due to the fact that there are three team members for.

**Comment by wilzbach** [comment]
@JackStouffer I was just about to reply, so here is my answer:

tl;dr: it was called `Final` in the [beginning](https://github.com/dlang/phobos/pull/2881#issue-54830087).

My take on this is that it behaves similarly to [final](https://en.wikipedia.org/wiki/Final_(Java)#Final_variables) variables in Java:

> A final variable can only be initialized once, either via an initializer or an assignment statement.

Okay it's a bit different as there's no language support:

> [..] It does not need to be initialized at the point of declaration: this is called a "blank final" variable. A blank final instance variable of a class must be definitely assigned in every constructor of the class in which it is declared

 For completion, the arguments against `Final` were:

@Dicebot:
> I also find Final confusing. My first thought reading the PR name was "how it is related to final methods?"

@ZombineDev:
> The only meaning that final has is virtual(false). Let's keep it that way.
[...]
> But, calling it Final would be inconsistent with the rest of the language! The only meaning that final has is in the context of inheritance, which is good. Just imagine exlaining to someone new to D that Final means head const and has nothing to do with final. Please keep such Java-isms out of D!

**Comment by ZombineDev** [comment]
`HeadConst` is a better name than `Final`. Apart from the argument about the confusion with `final`, I would really like if typecons offered symmetrical API. E.g 
HeadConst <> TailConst
Optional / Nullable <> Required / NotNull
HeadShared <> TailShared / HeadUnshared
Final <> Overridable
etc.
